tests/test_strat.py

```diff
@@ -101,18 +105,113 @@ def check_efficiency():
     print(f'Time elapse for duplicate: {time_elapse:.4f}')
 
 
-def test_adjoint():
-    sde = Ex4(d=d, m=m, sde_type=SDE_TYPES.stratonovich).to(device)
+def test_adjoint_inputs():
+    sde = problems.Ex4(d=d, m=m, sde_type=SDE_TYPES.stratonovich).to(device)
     bm = BrownianInterval(t0=t0, t1=t1, shape=(batch_size, m), dtype=dtype, device=device)
 
-    def func(y0):
+    def func(inputs, modules):
+        y0, sde = inputs[0], modules[0]
         ys = sdeint_adjoint(sde, y0, ts, bm, method='midpoint')
-        return ys[-1].sum()
+        return (ys[-1] ** 2).sum(dim=1).mean(dim=0)
 
-    y0_ = y0.clone().requires_grad_(True)
-    torch.autograd.gradcheck(func, y0_, rtol=1e-4, atol=1e-3, eps=1e-8)
+    swiss_knife_gradcheck(func, y0, sde, eps=1e-7, rtol=1e-3, atol=1e-3, grad_inputs=True, gradgrad_inputs=True)
+
+
+def test_adjoint_params():
+    sde = problems.Ex5(d=d, m=m, sde_type=SDE_TYPES.stratonovich).to(device)
+    bm = BrownianInterval(t0=t0, t1=t1, shape=(batch_size, m), dtype=dtype, device=device)
+
+    def func(inputs, modules):
+        """Outputs gradient norm squared."""
+        y0, sde = inputs[0], modules[0]
```

Comment:
So this is because `inputs` and `modules` are expected to be 1-tuples, right? Can there be asserts validating that this is the case?

Reply:
The input format is easy to check. However, it's hard to check that the input`func` has this type of input format. 

I noted this in the docstring.